### PR TITLE
Feature/fbeta integration

### PR DESCRIPTION
**Pull request by tehwenyi**
- Replaces `test.py` with `evaluate_coco.py` EDIT: Renamed `evaluate_coco.py` to `test_coco.py` instead
- Add evaluation of FBeta into the new `test.py`
- Add option to save results into `.csv` file
- Replace `attempt_load` function in `models/experimental.py` with `attempt_load_state_dict` function in `inference` branch

### CODE REVIEW

**Comment by tehwenyi** [comment]
Screenshot from 2023-09-21 15-41-03

Example output

**Comment by tehwenyi** [comment]
@yhsmiley made changes, do check! thank youuu!

File: models/experimental.py [hunk 1]
```diff
@@ -244,29 +244,36 @@ def forward(self, x):
 
 
 
-def attempt_load(weights, map_location=None):
```
Comment (yhsmiley) [inline models/experimental.py:247]: We should keep the `attempt_load` function also and not replace it, because there are other scripts that use it

**Review by yhsmiley — CHANGES_REQUESTED** [review]

**Comment by tehwenyi** [comment]
Removed Precision and Recall from the final line as it's unsure whether it's the Precision used to calculate the best F1 or F2 score. 
New output:

image

File: README.md [hunk 2]
```diff
@@ -27,6 +27,45 @@ python reparameterization.py --model_arch yolov7 --training_ckpt runs/train/yolo
 python save_state_dict.py --weights weights/yolov7_last.pt --save_path weights/yolov7_last_state.pt --cfg cfg/deploy/yolov7.yaml
 ```
 
+## Testing/Evaluating (with pycocotools)
+`test_coco.py` is a Python script for evaluating models with COCO ground truth annotations. Please follow the instructions below to use it effectively.
+
+### Prerequisites
+
+Before using `test.py`, ensure the following prerequisites are met:
```
Comment (yhsmiley) [inline README.md]: Should this be `test_coco.py`?

File: README.md [hunk 3]
```diff
@@ -27,6 +27,45 @@ python reparameterization.py --model_arch yolov7 --training_ckpt runs/train/yolo
...
+
+1. **`fdet-api`**: Download [fdet-api](https://github.com/yhsmiley/fdet-api/blob/master/PythonAPI/pycocotools/cocoeval.py).
+
+1. **COCO JSON File**: Provide a COCO JSON file containing ground truth annotations in the same folder as your image dataset. JSON image file names should match image file names.
+
+1. **Weight Reparameterization**: Note that you have to reparameterize the weights before testing (refer to [Reparameterization](#reparameterization-to-convert-weights-for-use-in-inference-branch).
+
+1. **COCO Annotation Format**: Annotations (e.g., `cat_id`, `ann_id`, `image_id`, etc.) must adhere to standard COCO format (start from 1).
+
+1. **`iscrowd` Annotation**: Ensure `iscrowd` exists in the ground truth annotations.
```
Comment (yhsmiley) [inline README.md]: The latest commit in `fdet-api` repo fixes this issue

File: test_coco.py [hunk 4]
```diff
@@ -266,15 +270,59 @@ def test(data,
...
 
     # Return results
     s = f"\n{len(list(save_dir.glob('labels/*.txt')))} labels saved to {save_dir / 'labels'}" if save_txt else ''
     print(f"Results saved to {save_dir}{s}")
     maps = np.zeros(nc) + map
     for i, c in enumerate(ap_class):
         maps[c] = ap[i]
+    if evaluate_fbeta_success:
+        return (f1_score, f2_score, mp, mr, map50, map, *(loss.cpu() / len(dataloader)).tolist()), maps, t
     return (mp, mr, map50, map, *(loss.cpu() / len(dataloader)).tolist()), maps, t
```
Comment (yhsmiley) [inline test_coco.py]: Hmm actually for this code the return function does nothing hor. Like the results are not used anywhere

File: test_coco.py [hunk 5]
```diff
@@ -266,15 +270,59 @@ def test(data,
...
+            if evaluate_fbeta:
+                cocoEval.accumulateFBeta()
+                cocoEval.summarizeFBetaScores(average=fbeta_average_method)
+                f1_score, _, _, _ = cocoEval.getBestFBeta(beta=1, iouThr=iou_thres, average=fbeta_average_method)
+                f2_score, _, _, _ = cocoEval.getBestFBeta(beta=2, iouThr=iou_thres, average=fbeta_average_method)
+                print(f"Results:")
+                s = ('%20s ' * 4) % ('f1-score', 'f2-score', 'map', 'map50')
+                print(f"{s}")
+                s = ('%20s ' * 4) % (f1_score, f2_score, map, map50)
+                print(f"{s}")
```
Comment (yhsmiley) [inline test_coco.py]: Hmm not sure if this would be confusing, because the values are calculated from different parameters

File: test_coco.py [hunk 6]
```diff
@@ -266,15 +270,59 @@ def test(data,
             cocoEval.accumulate()
             cocoEval.summarize()
             map, map50 = cocoEval.stats[:2]  # update results (mAP@0.5:0.95, mAP@0.5)
```
Comment (yhsmiley) [inline test_coco.py]: Omg i realise `map` is a bad variable name to use because its a keyword in python. Can shun bian change it pwease? :D

**Review by yhsmiley — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (tehwenyi) [reply README.md]: Yay nice! Ok will remove

**Comment by tehwenyi** [comment]
@yhsmiley done! do review
